In [1]:
import json
import pandas as pd

In [2]:
test = "test.json"
val = "val.json"

In [4]:
with open(test, "r") as f:
    test_data=json.load(f)
with open(val, "r") as f:
    val_data=json.load(f)

In [5]:
val_info_df = pd.DataFrame([val_data["info"]])  # single dict
val_audio_df = pd.DataFrame(val_data["audio"])
val_categories_df = pd.DataFrame(val_data["categories"])
val_annotations_df = pd.DataFrame(val_data["annotations"])
val_licenses_df = pd.DataFrame(val_data["licenses"])

test_info_df = pd.DataFrame([test_data["info"]])  # single dict
test_audio_df = pd.DataFrame(test_data["audio"])
test_categories_df = pd.DataFrame(test_data["categories"])
test_annotations_df = pd.DataFrame(test_data["annotations"])
test_licenses_df = pd.DataFrame(test_data["licenses"])

In [15]:
val_annotations_df

,id,audio_id,category_id
0,105125,105125,2964
1,105114,105114,2964
2,105122,105122,2964
3,105124,105124,2964
4,105126,105126,2964
...,...,...,...
45693,31107,31107,1160
45694,31104,31104,1160
45695,31103,31103,1160
45696,31106,31106,1160


In [20]:
annot_df = pd.read_csv("augmented_annot.csv")

/tmp/ipykernel_4038/3423486819.py:1: DtypeWarning: Columns (19) have mixed types. Specify dtype option on import or set low_memory=False.
  annot_df = pd.read_csv("augmented_annot.csv")


In [23]:
annot_df['id_cat'].describe()

count    243632.000000
mean       3402.470107
std        1164.062708
min        1395.000000
25%        2422.000000
50%        3427.000000
75%        4401.000000
max        5568.000000
Name: id_cat, dtype: float64

In [21]:
annot_df.head()

,Unnamed: 0,id_annot,audio_id,id_cat,name,common_name,supercategory,kingdom,phylum,class,order,family,genus,specific_epithet,evaluation,duration,file_name,supercategory_id,name_id,aug_type
0,0,156691,156691,3612,Zonotrichia albicollis,White-throated Sparrow,Aves,Animalia,Chordata,Aves,Passeriformes,Passerellidae,Zonotrichia,albicollis,1,4.296,train/03612_Animalia_Chordata_Aves_Passeriform...,1,2218,NaN
1,1,182163,182163,4361,Troglodytes aedon,House Wren,Aves,Animalia,Chordata,Aves,Passeriformes,Troglodytidae,Troglodytes,aedon,1,2.333,train/04361_Animalia_Chordata_Aves_Passeriform...,1,2967,NaN
2,3,61500,61500,2061,Eudynamys scolopaceus,Asian Koel,Aves,Animalia,Chordata,Aves,Cuculiformes,Cuculidae,Eudynamys,scolopaceus,1,18.795,train/02061_Animalia_Chordata_Aves_Cuculiforme...,1,667,NaN
3,4,224298,224298,5147,Otus scops,Eurasian Scops Owl,Aves,Animalia,Chordata,Aves,Strigiformes,Strigidae,Otus,scops,1,13.700,train/05147_Animalia_Chordata_Aves_Strigiforme...,1,3753,NaN
4,6,37989,37989,1399,Accipiter cooperii,Cooper's Hawk,Aves,Animalia,Chordata,Aves,Accipitriformes,Accipitridae,Accipiter,cooperii,1,34.339,train/01399_Animalia_Chordata_Aves_Accipitrifo...,1,5,NaN


In [22]:
val_annotations_df['category_id'].describe()

count    45698.000000
mean      2938.899317
std       1368.728068
min         50.000000
25%       1952.000000
50%       3057.000000
75%       3906.000000
max       5561.000000
Name: category_id, dtype: float64

In [29]:
test_annotations_df['category_id'].describe()

count    49527.000000
mean      2943.725725
std       1364.410823
min         50.000000
25%       1975.000000
50%       3057.000000
75%       3905.000000
max       5561.000000
Name: category_id, dtype: float64

In [30]:
categories_df

,id,name,common_name,supercategory,kingdom,phylum,class,order,family,genus,specific_epithet,supercategory_id
0,1,Accipiter atricapillus,American Goshawk,Aves,Animalia,Chordata,Aves,Accipitriformes,Accipitridae,Accipiter,atricapillus,1
1,2,Accipiter badius,Shikra,Aves,Animalia,Chordata,Aves,Accipitriformes,Accipitridae,Accipiter,badius,1
2,3,Accipiter bicolor,Bicolored Hawk,Aves,Animalia,Chordata,Aves,Accipitriformes,Accipitridae,Accipiter,bicolor,1
3,4,Accipiter cirrocephalus,Collared Sparrowhawk,Aves,Animalia,Chordata,Aves,Accipitriformes,Accipitridae,Accipiter,cirrocephalus,1
4,5,Accipiter cooperii,Cooper's Hawk,Aves,Animalia,Chordata,Aves,Accipitriformes,Accipitridae,Accipiter,cooperii,1
...,...,...,...,...,...,...,...,...,...,...,...,...
4169,4170,Crotalus pyrrhus,Southwestern Speckled Rattlesnake,Reptilia,Animalia,Chordata,Reptilia,Squamata,Viperidae,Crotalus,pyrrhus,2
4170,4171,Crotalus ruber,Red Diamond Rattlesnake,Reptilia,Animalia,Chordata,Reptilia,Squamata,Viperidae,Crotalus,ruber,2
4171,4172,Crotalus viridis,Prairie Rattlesnake,Reptilia,Animalia,Chordata,Reptilia,Squamata,Viperidae,Crotalus,viridis,2
4172,4173,Sistrurus catenatus,Massasauga,Reptilia,Animalia,Chordata,Reptilia,Squamata,Viperidae,Sistrurus,catenatus,2


In [40]:
val_filtered_annot = val_annotations_df[val_annotations_df['category_id'].isin(annot_df['id_cat'])]

In [58]:
test_filtered_annot = test_annotations_df[test_annotations_df['category_id'].isin(annot_df['id_cat'])]

In [45]:
val_filtered_annot['category_id'].nunique()

983

983 *is* the number of species belonging to Aves, Mammilia and Reptilia in val split. Thus, the val data has been filitered correctly.

In [62]:
annot_df[~annot_df['id_cat'].isin(test_filtered_annot['category_id'])]['id_cat'].nunique()

3191

The species gap between train and test/val split is very huge. So, I'm gonna make a stratified 20% split of the species that are in training data but not in val/test split and test on them at the end of the model trainng, by combing it with the test data.

# Mapping the Test/Val 

In [68]:
id_to_name = annot_df.set_index('id_cat')['name_id'].to_dict()
id_to_super = annot_df.set_index('id_cat')['supercategory_id'].to_dict()

In [70]:
val_filtered_annot['name_id'] = val_filtered_annot['category_id'].map(id_to_name)
val_filtered_annot['supercategory_id'] = val_filtered_annot['category_id'].map(id_to_super)

test_filtered_annot['name_id'] = test_filtered_annot['category_id'].map(id_to_name)
test_filtered_annot['supercategory_id'] = test_filtered_annot['category_id'].map(id_to_super)

/tmp/ipykernel_4038/2784466570.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  val_filtered_annot['name_id'] = val_filtered_annot['category_id'].map(id_to_name)
/tmp/ipykernel_4038/2784466570.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  val_filtered_annot['supercategory_id'] = val_filtered_annot['category_id'].map(id_to_super)
/tmp/ipykernel_4038/2784466570.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = v

In [71]:
val_filtered_annot.to_csv('val_annotation.csv', index=False)

In [74]:
from sklearn.model_selection import StratifiedKFold, train_test_split

In [76]:
not_in_val_df = annot_df[~annot_df['id_cat'].isin(test_filtered_annot['category_id'])]
train, test = train_test_split(not_in_val_df, stratify=not_in_val_df['id_cat'], shuffle=True, random_state=42)

In [77]:
test.to_csv('rare_test.csv', index=True)

In [79]:
annot_df.drop(test.index, inplace=True)

In [80]:
annot_df.to_csv('augmented_annot.csv')

In [92]:
skf = StratifiedKFold(n_splits=3, shuffle=True)

In [83]:
test_x = test_filtered_annot.copy()
test_y = test_x.pop('name_id')

In [93]:
for i, (_, test_idx) in enumerate(skf.split(test_x, test_y)):
    fold = test_filtered_annot.iloc[test_idx].reset_index(drop=True)
    fold.to_csv(f'test_split_{i}.csv', index=False)

In [94]:
test_0 = pd.read_csv('test_split_0.csv')
test_1 = pd.read_csv('test_split_1.csv')
test_2 = pd.read_csv('test_split_2.csv')

In [95]:
test_0

,id,audio_id,category_id,name_id,supercategory_id
0,105056,105056,2964,1570,1
1,105070,105070,2964,1570,1
2,105065,105065,2964,1570,1
3,105074,105074,2964,1570,1
4,105061,105061,2964,1570,1
...,...,...,...,...,...
14042,51722,51722,1818,424,1
14043,51723,51723,1818,424,1
14044,70431,70431,2355,961,1
14045,70421,70421,2355,961,1
